### DISCUSSION TITLE
Unable to Load Abstract Single Table Inheritance Entities After Upgrade to Play-Ebean 8.3.0

### DISCUSSION THREAD

**Discussion by robby-macphail** [ebean-orm/ebean#3411] [General]
Perhaps this question is better suited for a Play-Ebean discussion group, but I am upgrading a Java Play application to Play-Ebean 8.30 from 7.30, both of which, I believe, use the same Ebean version (15.1.0). The application has many abstract entity classes using single table inheritance strategy along with a discriminator. This pattern has worked well previously, but since the upgrade, Ebean is unable to load any of the entities inheriting the abstract class.

For example:

@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "my_type_field", discriminatorType = DiscriminatorType.STRING)
public abstract class MyAbstractClass extends AnotherAbstractClass { ... }

@Entity
@DiscriminatorValue("FOO")
public class MyConcreteClass extends MyAbstractClass { ... }

Attempting to load the entity results in:

Caused by: jakarta.persistence.PersistenceException: java.lang.UnsupportedOperationException: cannot create entity bean for abstract entity MyAbstractClass
	at io.ebeaninternal.server.deploy.BeanDescriptor.createReference(BeanDescriptor.java:1521)
	at io.ebeaninternal.server.deploy.BeanDescriptor.contextRef(BeanDescriptor.java:1730)
	at io.ebeaninternal.server.deploy.AssocOneHelp.read(AssocOneHelp.java:70)
	at io.ebeaninternal.server.deploy.AssocOneHelp.readSet(AssocOneHelp.java:81)
	at io.ebeaninternal.server.deploy.BeanPropertyAssocOne.readSet(BeanPropertyAssocOne.java:624)
	at io.ebeaninternal.server.query.SqlBeanLoad.load(SqlBeanLoad.java:63)

Has anything changed in recent Ebean versions to disallow using abstract classes with the single table inheritance?

**Comment by rob-bygrave** [comment]
> use the same Ebean version (15.1.0).

You can _NOT_ use Ebean 15.x and _MUST_ use Ebean 14.x to use single table inheritance.

Ebean 14.x and 15.x versions are aligned with bug fixes with the difference being that Ebean 15.x has some features removed and that includes removing support for single table inheritance. So you need to use ebean 14.x (and that version is maintained and 14.x is still the master branch).

For background refer to:
https://github.com/ebean-orm/ebean/releases/tag/15.0.0-RC
https://github.com/ebean-orm/ebean/discussions/3112

So change your dependency from `15.1.0` to `14.1.0` and stay on 14.x releases of ebean.

**Reply by robby-macphail** [reply]
Thanks for the info. We're trying to upgrade to Play 3, so that presents a bit of a dilemma for us. The play-ebean plugin moved to version 15 from 13 starting with version 8.2.0 of the plugin. So unless Play plans to release a plugin that supports EBean 14, we're stuck on version 13. Our other option is to refactor our applications to remove all instances of single table inheritance, which is a significant piece or work.

**Reply by rbygrave** [reply]
> So unless Play plans to release a plugin that supports EBean 14, we're stuck on version 13

No in that Ebean 14 is the same as Ebean 15 but with more features. So you literally just need to explicitly use the Ebean 14.x dependencies.

Use the latest 14.x which is currently `14.3.0` - there was a bug in `14.1.0`.

The point of keeping 14.x, 14.x-javax and 15.x all in sync and releasing all three of them for each and every release is to avoid that dilemma and allow people time to upgrade from javax -> jakarta and upgrade from 14.x -> 15.x in terms of the features removed in 15.x.  I mean, you can use 13 but that isn't getting bug fixes etc and instead people are expected to get to 14.x or 14.x-javax.
